pkg/auth/oauth2/auth_openshift.go

```diff
@@ -221,18 +226,93 @@ func (o *openShiftAuth) LogoutRedirectURL() string {
 	return o.logoutRedirectOverride
 }
 
+func (o *openShiftAuth) reviewToken(token string) (*authv1.TokenReview, error) {
+	tokenReviewURL, err := url.Parse(o.issuerURL)
+	if err != nil {
+		return nil, err
+	}
+	tokenReviewURL.Path = tokenReviewPath
+
+	tokenReview := &authv1.TokenReview{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "authentication.k8s.io/v1",
+			Kind:       "TokenReview",
+		},
+		Spec: authv1.TokenReviewSpec{
+			Token: token,
+		},
+	}
+
+	tokenReviewJSON, err := json.Marshal(tokenReview)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, tokenReviewURL.String(), bytes.NewBuffer(tokenReviewJSON))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", o.internalK8sConfig.BearerToken))
+
+	res, err := o.k8sClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusCreated {
+		return nil, fmt.Errorf("unable to validate user token: %v", res.Status)
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal the response into a TokenReview object
+	var responseTokenReview authv1.TokenReview
+	err = json.Unmarshal(body, &responseTokenReview)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if the token is authenticated
+	if !responseTokenReview.Status.Authenticated {
+		err := fmt.Errorf("invalid token: %v", token)
+		if responseTokenReview.Status.Error != "" {
+			err = fmt.Errorf("invalid token: %s", responseTokenReview.Status.Error)
+		}
+		return nil, err
+	}
+
+	return tokenReview, nil
+}
+
 func (o *openShiftAuth) Authenticate(_ http.ResponseWriter, r *http.Request) (*auth.User, error) {
-	// TODO: This doesn't do any validation of the cookie with the assumption that the
-	// API server will reject tokens it doesn't recognize. If we want to keep some backend
-	// state we should sign this cookie. If not there's not much we can do.
 	cookie, err := r.Cookie(sessions.OpenshiftAccessTokenCookieName)
 	if err != nil {
 		return nil, err
 	}
+
 	if cookie.Value == "" {
 		return nil, fmt.Errorf("unauthenticated, no value for cookie %s", sessions.OpenshiftAccessTokenCookieName)
 	}
 
+	if o.internalK8sConfig.BearerToken != "" {
```

Comment:
I'm assuming this check is to make off-cluster work? This is always set when running in-cluster, correct?

Reply:
Yes, if we don't do this check, it breaks in off-cluster mode because the internal K8s config is unset, which means the TokenReview request fires off with no bearer token and always fails.

Reply:
I think this is safe since if there is no service account bearer token there is no risk of privilege escalation. All requests have to use the user's token because it's all that we have.